taroscript/tx.go

```diff
@@ -327,35 +329,123 @@ func InputScriptSpendSigHash(virtualTx *wire.MsgTx, input *asset.Asset,
+	// time? What are the performance and security implications?
+	if len(vIn.TaprootBip32Derivation) > 1 {
+		return nil, fmt.Errorf("unsupported multiple taproot BIP32 " +
+			"derivation info found, can only sign for one at a" +
+			"time")
+	}
+	if len(vIn.TaprootBip32Derivation[0].LeafHashes) > 1 {
+		return nil, fmt.Errorf("unsupported number of leaf hashes in " +
+			"taproot BIP32 derivation info, can only sign for one " +
+			"at a time")
+	}
+
+	derivation := vIn.TaprootBip32Derivation[0]
 
 	// Compute a virtual prevOut from the input asset for the signer.
-	prevOut, err := InputAssetPrevOut(*inputAsset)
+	prevOut, err := InputAssetPrevOut(*vIn.Asset())
 	if err != nil {
 		return nil, err
 	}
 
-	// Build the signing descriptor for a BIP 86 signature over the Taro
-	// virtual TX. To follow BIP 86, no tweaks nor witness script are
-	// provided, and the untweaked internal key is used as the signing key.
+	// Start with a default sign descriptor and the BIP 86 sign method then
+	// adjust depending on the input parameters.
 	spendDesc := lndclient.SignDescriptor{
 		KeyDesc: keychain.KeyDescriptor{
-			PubKey: &internalKey,
+			PubKey: vIn.Asset().ScriptKey.RawKey.PubKey,
 		},
 		SignMethod: input.TaprootKeySpendBIP0086SignMethod,
 		Output:     prevOut,
-		HashType:   sigHashType,
+		HashType:   vIn.SighashType,
 		InputIndex: idx,
 	}
 
+	// There are three possible signing cases: BIP 86 key spend path, key
+	// spend path with a script root, and script spend path.
+	switch {
+	// If there is no merkle root, we're doing a BIP 86 key spend.
+	case len(vIn.TaprootMerkleRoot) == 0:
+		// This is the default case, so we don't need to do anything.
+
+	// No leaf hash means we're not signing a specific script, so this is
+	// the key spend path with a script root.
+	case len(vIn.TaprootMerkleRoot) == sha256.Size &&
+		len(derivation.LeafHashes) == 0:
+
+		spendDesc.SignMethod = input.TaprootKeySpendSignMethod
+		spendDesc.TapTweak = vIn.TaprootMerkleRoot
+
+	// One leaf hash and a merkle root means we're signing a specific
+	// script.
+	case len(vIn.TaprootMerkleRoot) == sha256.Size &&
+		len(derivation.LeafHashes) == 1:
+
+		// If we're supposed to be signing for a leaf hash, we also
+		// expect the leaf script that hashes to that hash in the
+		// appropriate field.
+		if len(vIn.TaprootLeafScript) != 1 {
+			return nil, fmt.Errorf("specified leaf hash in " +
+				"taproot BIP32 derivation but missing taproot " +
+				"leaf script")
+		}
+
+		leafScript := vIn.TaprootLeafScript[0]
+		leaf := txscript.TapLeaf{
+			LeafVersion: leafScript.LeafVersion,
+			Script:      leafScript.Script,
+		}
+		leafHash := leaf.TapHash()
+		if !bytes.Equal(leafHash[:], derivation.LeafHashes[0]) {
+			return nil, fmt.Errorf("specified leaf hash in " +
+				"taproot BIP32 derivation but corresponding " +
+				"taproot leaf script was not found")
+		}
+
+		spendDesc.SignMethod = input.TaprootScriptSpendSignMethod
+		spendDesc.TapTweak = vIn.TaprootMerkleRoot
```

Comment:
Why is the tap tweak needed for a script spend from the tree? This is only needed for keyspend paths.

Reply:
Good point, removed.